### DISCUSSION TITLE
Support Apollo Cache Control Specification

### DISCUSSION THREAD

**Discussion by ttyniwa** [ExpediaGroup/graphql-kotlin#1315] [Ideas] [answered]
# Support [Apollo Cache Control Specification](https://github.com/apollographql/apollo-cache-control)

Apollo Cache Control exposes cache control hints for an individual request under a cacheControl key in extensions.
graphql-java already implements this feature as [CacheControl](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/cachecontrol/CacheControl.java) class.
I'd like to use this feature, so can I implement it?

## code

This can be easily implemented by modifying [ line 45 of GraphQLRequestHandler ](https://github.com/ExpediaGroup/graphql-kotlin/blob/master/servers/graphql-kotlin-server/src/main/kotlin/com/expediagroup/graphql/server/execution/GraphQLRequestHandler.kt#L45) as follows.
```
+            var executionResult = graphQL.executeAsync(executionInput).await()
+            executionResult = executionInput.cacheControl.addTo(executionResult)
+            executionResult.toGraphQLResponse()
```

## How to use

Call below code at resolver.
```
    fun books(
        env: DataFetchingEnvironment,
    ): List<Content> {
+        env.cacheControl.hint(env, 60, CacheControl.Scope.PUBLIC)
...
```

## response sample
```
  "extensions": {
    "cacheControl": {
      "version": 1,
      "hints": [
        {
          "path": [
            "books"
          ],
          "maxAge": 60,
          "scope": "PUBLIC"
        }
      ]
    }
  }
```

## Caution

If any resolver doesn't call the cacheControl.hint() function, the cacheControl will still appear in the response as following.

```
  "extensions": {
    "cacheControl": {
      "version": 1,
      "hints": []
    }
  }
```

This is a specification of CacheControl in graphql-java.
If we don't want them to appear, we need to modify the CacheControl class.

**Comment by dariuszkuc** [comment, marked as answer]
Hello 👋 
`GraphQLRequestHandler` is an open class so you can easily extend it and implement the cache control logic above.

**Reply by ttyniwa** [reply]
Thanks for reply.
Graphql-java provides the CacheControl impl, and DataFetchingEnvironment.cacheControl is also initialized by default.
Therefore, i think it would be better if graphql-kotlin also supports this feature by default.

**Reply by dariuszkuc** [reply]
`Apollo Cache Control` as the name implies is specific to Apollo and not part of the GraphQL spec. While it is pretty popular it should not be the global default.
